### DISCUSSION TITLE
Allow giving named workspaces a fixed index, or not giving them an index entirely

### DISCUSSION THREAD

**Discussion by maxverbeek** [YaLTeR/niri#805] [Ideas]
Hi!

I've switched to Niri a little while ago, and started tinkering with an Astal (formerly Ags) config for it. One of the workflows I'm trying to create is the following:

Pin some programs, such as my browser, slack, spotify, etc... to some keybinds (say Mod+8, Mod+9, Mod+0). I can do this by creating named workspaces, assign above keybinds to focus these workspaces, and call it a day. Everytime I press those keybinds I want to focus the correct workspace on the correct monitor, without having to navigate to that monitor beforehand (named workspaces solve this).

I've noticed that these named workspaces occupy some index in Niri's layout. What I want to achieve is that the first half of my number row (1-5) maps to a dynamic workspace (per monitor) and the second half of my number row (6-0) to "fixed" workspaces, which will live on a specific monitor, so that when I want to navigate to one of these windows I do not have to first navigate to the correct monitor.

I'm not sure if this can already be done. The problem that I ran into is, when creating named workspaces they occupy an index (starting from 1) on the monitor they are created on. This means that my dynamic workspace indices start after my named workspace index, on a fairly arbitrary number. This becomes worse when moving my named workspaces around, changing the index of all of my dynamic workspaces.

A fairly neat way (configuration wise, probably not implementation wise) is if we can set an index for named workspaces:

```kdl
workspace "browser" {
    open-on-output "DP-1"
    index 10 # hotkey 0
}
```

I'm not entirely sure what the behaviour is when moving this workspace to other monitors (where a workspace on index 10 might already exist), or how this affects moving to the next (index 11) and previous (first index lower than 10?) workspace.

A more constrained way is probably to allow workspaces to not have an index at all, and have them not interfere with any sort of index (not sure on the configuration syntax here)

```kdl
workspace "browser" {
    open-on-output "DP-1"
    index false
}
```

making it so that this workspace can only be reached through `Action::FocusWorkspace(Name("browser"))` keybinds. Again, not exactly sure what the behaviour is regarding the relation to previous/next workspaces. Maybe they do not exist relative to this workspace (similar to how there is no workspace above index 0).

What are your thoughts on this? If it reasonably fits in the architecture of Niri's layout code I am willing to give the implementation a shot myself.

Do you think the workflow I describe is already possible to implement somehow? If so I would really like to know :sweat_smile: 

Cheers!

**Comment by YaLTeR** [comment]
Hey, so yeah, named workspaces intentionally behave just like normal workspaces (you can move them around, etc.) except only for sticking around without windows, and being able to focus them by name. So you will always be able to move something into slots 1-5 and have indexing point there.

You might be looking for something like scratch workspaces as described here: https://github.com/YaLTeR/niri/discussions/329#discussioncomment-11219412 They don't show up as normal workspaces, and when you focus them, they overlay on top of the current workspace. I think this would be better than overriding indexing behavior.

**Reply by maxverbeek** [reply]
Thanks! Yeah I think something like that will work for what I had in mind.

**Reply by kcxt** [reply]
Hiya, sorry to necropost but just chiming in with my own experience: I've been using Niri for a few months now on my laptop and it has been fantastic overall.

On sway I mostly treated the first 10 workspaces as "dynamic" in that I would just use them for anything, but I usually have a few fixed workspaces for chat apps, web browsing and email. On niri I was just using the first workspace for this but I didn't find that very intuitive since the first workspace (`Mod+1`) is pretty ingrained as the one where I'm usually working.

With that in mind, I found it a bit annoying when I tried to define a fixed "chat" workspace, configured a keybind for it (`Mod+'` on my british keyboard) and then found that `Mod+1` still takes me to it.

Maybe this is arbitrary but what if fixed workspaces just started their indexes from 100? This would always put them at the end of the dynamic workspaces and while it's a maybe weird rule to learn it's simple to apply.

Alternatively and maybe more practically I'd agree with @maxverbeek's originally proposal to let you set an index for a fixed workspace.

Although I can see a scratchpad being an ok solution as well, being "inserted" on top of or below the current workspace so it's easy to switch back with trackpad gestures.

**Reply by marcusxss** [reply]
I was just about to open a discussion about this but then found this one.

For scratchpads it would help to be able to create them without an index IMO, so instead one calls applications back and forth from there and not visit the workspace directly, or alternatively it IS visited directly but via a specific, predefined keybind not used by the dynamically created, indexed workspaces.

The alternative I've found is switching entirely away from dynamic workspaces and only using named ones, but that feels like quite the sacrifice.

**Reply by gardenerik** [reply]
The workaround I have used is that I live with the fact that named workspaces are at the beginning of the index space, but I have changed "Mod+1" to focus first unnamed workspace (in my case, I have two named ones, so that would be workspace 3, Mod+2 then focuses 4, etc.).

Edit: I have just found out that this only really works for single monitor setups.

**Reply by appleseedexm** [reply]
I'm glad I'm not the only one that uses niri like this. 
I have a workflow where I want to find a set of applications on a certain workspace, dynamic indexing makes this very hard. I now have hard keybinds and named workspaces for 1-8, without using index, but I wished I could move from 1 to 5 with `Mod+J` if 2-4 had no content (since they are named workspaces, they exist as empty).
If I could I'd also lock that these named workspaces always have to be in this particular order, or even put them in a block where dynamic workspaces have to move around the whole block.
